pkg/daemon/ceph/client/upgrade.go

```diff
@@ -104,6 +147,193 @@ func EnableNautilusOSD(context *clusterd.Context) error {
+			// now looping over a single element since we can't address the key directly (we don't know its name)
+			for _, monCount := range versions.Mon {
+				if monCount < 3 {
+					logger.Infof("the cluster has less than 3 monitors, not performing upgrade check, running in best-effort")
+					return nil
+				}
+			}
+		}
+	// Trying to handle the case where a **single** osd is deployed and an upgrade is called
+	case "osd":
+		// if len(versions.Osd) > 1, we have different Ceph versions for some osd(s)
+		// this is fine running the upgrade we can run the upgrade checks
+		if len(versions.Osd) == 1 {
+			// now trying to parse and find how many osds are presents
+			// if we have less than 3 osds we skip the check and do best-effort
+			for _, osdCount := range versions.Osd {
+				if osdCount < 3 {
+					logger.Infof("the cluster has less than 3 OSDs, not performing upgrade check, running in best-effort")
+					return nil
+				}
+			}
+		}
+	}
+	// we don't implement any checks for mon, rgw and rbdmirror since:
+	//  - mon: the is done in the monitor code since it ensures all the mons are always in quorum before continuing
+	//  - rgw: the pod spec has a liveness probe so if the pod successfully start
+	//  - rbdmirror: you can chain as many as you want like mdss but there is no ok-to-stop logic yet
+	err = okToStopDaemon(context, deployment)
+	if err != nil {
+		return fmt.Errorf("failed to check if %s was ok to stop. %+v", deployment, err)
+	}
+
+	return nil
+}
+
+// OkToContinue determines if it's ok to continue an upgrade
+func OkToContinue(context *clusterd.Context, namespace, deployment string, cephVersion cephver.CephVersion) error {
+	daemonName, err := findDaemonName(deployment)
+	if err != nil {
+		logger.Warningf("%+v", err)
+		return nil
+	}
+
+	// the mon case is handled directly in the deployment where the mon checks for quorum
+	switch daemonName {
+	case "osd":
+		err := okToContinueOSDDaemon(context, namespace)
+		if err != nil {
+			return fmt.Errorf("failed to check if %s was ok to continue. %+v", deployment, err)
+		}
+	case "mds":
+		err := okToContinueMDSDaemon(context, namespace, deployment)
+		if err != nil {
+			return fmt.Errorf("failed to check if %s was ok to continue. %+v", deployment, err)
+		}
+	}
+
+	return nil
+}
+
+func okToStopDaemon(context *clusterd.Context, deployment string) error {
+	daemonID := findDaemonID(deployment)
+	daemonName, err := findDaemonName(deployment)
+	if err != nil {
+		logger.Warningf("%+v", err)
+		return nil
+	}
+
+	if !stringInSlice(daemonName, daemonNoCheck) {
+		output, err := context.Executor.ExecuteCommandWithOutput(false, "", "ceph", daemonName, "ok-to-stop", daemonID)
+		if err != nil {
+			return fmt.Errorf("deployment %s cannot be stopped. %+v", deployment, err)
+		}
+		logger.Infof("deployment %s is ok to be updated. %s", deployment, output)
+	}
+
+	logger.Infof("deployment %s is ok to be updated.", deployment)
```

Comment:
How about debug messages for a lot of these checks in the ok to stop methods? If there are a lot of daemons, this would be verbose.